File: ios/iosApp/Views/MiniAppDetails/MiniAppDetailsViewModel.swift
import Foundation

@MainActor
class MiniAppDetailsViewModel: ObservableObject {
    @Published var isDownloaded: Bool = false
    @Published var isLoading: Bool = false
    @Published var error: String? = nil

    @Published var launchPath: String? = nil
    @Published var launchBundle: Bundle_? = nil
    @Published var launchManifest: WebRuntimeManifest? = nil
    @Published var isShowingRuntime: Bool = false

    private let downloadBundleUseCase: DownloadBundleUseCase
    private let fileManager = FileManager.default

    init(downloadBundleUseCase: DownloadBundleUseCase) {
        self.downloadBundleUseCase = downloadBundleUseCase
    }

    func checkDownloadStatus(bundle: Bundle_) {
        let documents = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first!
        let bundlePath = documents.appendingPathComponent("bundles").appendingPathComponent(bundle.id)
        let manifestPath = bundlePath.appendingPathComponent("manifest.json")
        isDownloaded = fileManager.fileExists(atPath: manifestPath.path)
    }

    func downloadAndLaunch(bundle: Bundle_) async {
        isLoading = true
        error = nil

        let result = await downloadBundleUseCase.execute(bundleId: bundle.id)

        switch result {
        case .success(let path):
            launchPath = path
            launchBundle = bundle
            isDownloaded = true
            parseManifestAndLaunch(path: path, bundle: bundle)
        case .error(let appError):
            error = message(from: appError)
        }

        isLoading = false
    }

    func openMiniApp(bundle: Bundle_) {
        let documents = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first!
        let bundlePath = documents.appendingPathComponent("bundles").appendingPathComponent(bundle.id)

        if fileManager.fileExists(atPath: bundlePath.path) {
            launchPath = bundlePath.path
            launchBundle = bundle
            parseManifestAndLaunch(path: bundlePath.path, bundle: bundle)
        } else {
            Task { await downloadAndLaunch(bundle: bundle) }
        }
    }

    func uninstallBundle(bundle: Bundle_) {
        let documents = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first!
        let bundlePath = documents.appendingPathComponent("bundles").appendingPathComponent(bundle.id)

        do {
            if fileManager.fileExists(atPath: bundlePath.path) {
                try fileManager.removeItem(at: bundlePath)
                isDownloaded = false
            }
        } catch {
            self.error = "Failed to uninstall: \(error.localizedDescription)"
        }
    }

    private func parseManifestAndLaunch(path: String, bundle: Bundle_) {
        let bundleUrl = URL(fileURLWithPath: path)
        let manifestUrl = bundleUrl.appendingPathComponent("manifest.json")

        guard
            let data = try? Data(contentsOf: manifestUrl),
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            self.error = "Invalid manifest.json"
            return
        }

        let id = json["id"] as? String ?? bundle.id
        let version = json["version"] as? String ?? String(json["version"] as? Int ?? 1)
        let name = json["name"] as? String ?? bundle.name
        let entry = json["entry"] as? String ?? "index.html"
        let typeString = (json["type"] as? String)?.uppercased() ?? "STANDARD"
        let orientation = json["orientation"] as? String ?? "portrait"
        let fullScreen = json["fullScreen"] as? Bool ?? true
        let runtimeType = typeString == "FLUTTER" ? "flutter" : "standard"

        self.launchManifest = WebRuntimeManifest(
            id: id,
            version: version,
            name: name,
            entry: entry,
            type: runtimeType,
            orientation: orientation,
            fullScreen: fullScreen
        )

        self.isShowingRuntime = true
    }

    private func message(from error: AppError) -> String {
        switch error {
        case .networkError(let message, _): return message ?? "Network error"
        case .serverError(_, let message): return message ?? "Server error"
        case .databaseError(let message, _): return message ?? "Database error"
        case .unknownError(let message, _): return message ?? "Unknown error"
        case .validationError(let message): return message ?? "Validation error"
        }
    }
}

File: android/app/src/main/kotlin/com/lepos/lepos/ui/miniappdetails/MiniAppDetailsViewModel.kt
package com.lepos.lepos.ui.miniappdetails

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.lepos.lepos.domain.model.RuntimeType
import com.lepos.lepos.domain.model.WebRuntimeManifest
import com.lepos.lepos.domain.model.bundle.*
import com.lepos.lepos.domain.usecase.*
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import org.json.JSONObject
import java.io.File

class MiniAppDetailsViewModel(
    private val downloadBundleUseCase: DownloadBundleUseCase,
    private val getBundleStatsUseCase: GetBundleStatsUseCase,
    private val getBundlePromotionsUseCase: GetBundlePromotionsUseCase,
    private val trackBundleDownloadUseCase: TrackBundleDownloadUseCase,
    private val context: Context
) : ViewModel() {

    private val _isDownloaded = MutableStateFlow(false)
    val isDownloaded: StateFlow<Boolean> = _isDownloaded.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _bundleStats = MutableStateFlow<BundleStats?>(null)
    val bundleStats: StateFlow<BundleStats?> = _bundleStats.asStateFlow()

    private val _promotions = MutableStateFlow<List<BundlePromotion>>(emptyList())
    val promotions: StateFlow<List<BundlePromotion>> = _promotions.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private val _launchManifest = MutableStateFlow<WebRuntimeManifest?>(null)
    val launchManifest: StateFlow<WebRuntimeManifest?> = _launchManifest.asStateFlow()

    fun loadBundleDetails(bundleId: String) {
        viewModelScope.launch {
            _isLoading.value = true
            
            // Parallel fetching
            val statsDeferred = async { getBundleStatsUseCase(bundleId) }
            val promotionsDeferred = async { getBundlePromotionsUseCase(bundleId) }
            
            val statsResult = statsDeferred.await()
            val promotionsResult = promotionsDeferred.await()
            
            if (statsResult is com.lepos.lepos.core.Result.Success) {
                _bundleStats.value = statsResult.data
            }
            
            if (promotionsResult is com.lepos.lepos.core.Result.Success) {
                _promotions.value = promotionsResult.data
            }
            
            _isLoading.value = false
        }
    }

    fun checkDownloadStatus(bundle: Bundle) {
        val miniAppsDir = File(context.filesDir, "mini-apps")
        val bundleDir = File(miniAppsDir, bundle.id)
        val manifestFile = File(bundleDir, "manifest.json")
        _isDownloaded.value = manifestFile.exists()
    }

    fun downloadAndOpen(bundle: Bundle) {
        if (_isLoading.value) return

        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null

            downloadBundleUseCase.invoke(bundle)
                .onSuccess { path ->
                    _isDownloaded.value = true
                    parseManifestAndLaunch(path, bundle)
                    
                    // Track download in background
                    viewModelScope.launch {
                        trackBundleDownloadUseCase.invoke(bundle.id)
                    }
                    
                    _isLoading.value = false
                }
                .onError { error ->
                    _error.value = error.message
                    _isLoading.value = false
                }
        }
    }

    fun openMiniApp(bundle: Bundle) {
        val miniAppsDir = File(context.filesDir, "mini-apps")
        val bundleDir = File(miniAppsDir, bundle.id)

        if (bundleDir.exists()) {
            parseManifestAndLaunch(bundleDir.absolutePath, bundle)
        } else {
            // Re-download if missing
            downloadAndOpen(bundle)
        }
    }

    fun uninstallBundle(bundle: Bundle) {
        try {
            val miniAppsDir = File(context.filesDir, "mini-apps")
            val bundleDir = File(miniAppsDir, bundle.id)
            if (bundleDir.exists()) {
                bundleDir.deleteRecursively()
                _isDownloaded.value = false
            }
        } catch (e: Exception) {
            _error.value = "Uninstall failed: ${e.message}"
        }
    }

    private fun parseManifestAndLaunch(path: String, bundle: Bundle) {
        try {
            val manifestFile = File(path, "manifest.json")
            if (manifestFile.exists()) {
                val content = manifestFile.readText()
                val json = JSONObject(content)

                val id = json.optString("id", bundle.id)
                val version = json.optString("version", "1")
                val name = json.optString("name", bundle.name)
                val entry = json.optString("entry", "index.html")
                val typeString = json.optString("type", "STANDARD").uppercase()
                val orientation = json.optString("orientation", "portrait")
                val fullScreen = json.optBoolean("fullScreen", true)

                val runtimeType =
                    if (typeString == "FLUTTER") RuntimeType.FLUTTER else RuntimeType.STANDARD

                val manifest = WebRuntimeManifest(
                    id = id,
                    version = version,
                    name = name,
                    entry = entry,
                    type = runtimeType,
                    orientation = orientation,
                    fullScreen = fullScreen
                )
                _launchManifest.value = manifest
            }
        } catch (e: Exception) {
            _error.value = "Invalid manifest: ${e.message}"
        }
    }

    fun clearLaunchManifest() {
        _launchManifest.value = null
    }
}
